File: package.py
from main import app
import os
import json
import time
import util
from cfapi import REPOSITORY

class Package:
    id: int = -1
    name: str = ""
    description: str = ""
    repository: str = app.config["REPO_URL"]
    repository_id: int = -1
    releases: list = []
    authors: list = []
    slug: str = ""

    def __init__(self, id, name, slug, description, authors, repository_id, repository=app.config["REPO_URL"], releases = []):
        self.id = id
        self.name = name
        self.slug = slug
        self.description = description
        self.authors = authors
        self.repository = repository
        self.repository_id = repository_id
        self.releases = releases
        
    @classmethod
    def create_new(cls, name, slug, description, authors, repository=app.config['REPO_URL'], repository_id=None):
        f = open("./assets/pkg_index.json", "r")
        j = util.get_idx()
        if repository_id:
            print(repository_id)
            p = cls(
                id=j["next_pkg_id"], 
                name=name,
                slug=slug,
                description=description,
                authors=authors,
                repository=repository,
                repository_id=repository_id
                )
        else:
            p = cls(
            id=j["next_pkg_id"], 
            name=name, 
            slug=slug,
            description=description,
            authors=authors,
            repository=repository,
            repository_id=j["next_pkg_id"]
            )
        # os.mkdir(f"./assets/{j['next_pkg_id']}")
        j["next_pkg_id"] += 1
        j["packages"].append(p.as_json())
        f = open("./assets/pkg_index.json", "w")
        f.write(json.dumps(j))
        f.close()
        
        return p

    @classmethod
    def obj_from_id(cls, id: int):
        f = open("./assets/pkg_index.json", "r")
        j = json.loads(f.read())
        f.close()
        p_json = j['packages'][id]
        p = cls(
            id=p_json['id'],
            name=p_json['name'],
            slug=p_json['slug'],
            description=p_json['description'],
            authors=p_json['authors'],
            repository=p_json['repository'],
            repository_id=p_json['repository_id'],
            releases=p_json['releases']
        )
        return p

    def as_json(self):
        j = json.loads('{}')
        j['id'] = self.id
        j['name'] = self.name
        j['slug'] = self.slug
        j['description'] = self.description
        j['authors'] = self.authors
        j['repository'] = self.repository
        j['repository_id'] = self.repository_id
        j['releases'] = self.releases
        return j
    
    def next_release_id(self):
        return len(self.releases)

class Release:
    id: int = 1
    version: str = ""
    game_version: str = ""
    deps: list = []
    parent_package_id: int = 1
    released: int = 0
    downloads: int = 0
    direct_link: str = ""
    prefer_link: bool = False

    def __init__(self, id, version, game_version, deps: list, parent_package_id: int, direct_link: str, prefer_link: bool = False, released=time.time_ns()):
        self.id = id
        self.version = version
        self.game_version = game_version
        self.deps = deps
        self.parent_package_id = parent_package_id
        self.released = released
        self.direct_link = direct_link
        self.prefer_link = prefer_link
    
    @classmethod
    def create_new(cls, version, game_version, deps: list, parent_package_id: int, direct_link = None, prefer_link: bool = False, released=time.time_ns()):
        f = open("./assets/pkg_index.json")
        j = json.loads(f.read())

        this_id = len(j['packages'][parent_package_id]['releases'])

        if direct_link:
            dlink = direct_link
        else:
            dlink = app.config["REPO_URL"] + f"/v1/download_release/{parent_package_id}/{this_id}"

        rel = cls(
            id=this_id,
            version=version,
            game_version=game_version,
            deps=deps,
            parent_package_id=parent_package_id,
            direct_link=dlink,
            prefer_link=prefer_link,
            released=released
        )

        j['packages'][parent_package_id]['releases'].append(rel.as_json())
        f = open("./assets/pkg_index.json", "w")
        f.write(json.dumps(j))
        f.close()
        # os.mkdir(f"./assets/{parent_package_id}/{this_id}")
    
    @classmethod 
    def obj_from_id(cls, pkg_id, release_id):
        f = open("./assets/pkg_index.json", "r")
        j = json.loads(f.read())

        rel_json = j['packages'][pkg_id]['releases'][release_id]

        rel = cls(
            id = rel_json['id'],
            version = rel_json['version'],
            game_version = rel_json['game_version'],
            deps = rel_json['deps'],
            parent_package_id = rel_json['parent_package_id'],
            released = rel_json['released'],
            direct_link=rel_json['direct_link'],
            prefer_link = j['prefer_link']
        )

    def as_json(self):
        j = json.loads('{}')
        j['id'] = self.id
        j['version'] = self.version
        j['game_version'] = self.game_version
        # j['deps'] = []
        j['deps'] = self.deps
        j['parent_package_id'] = self.parent_package_id
        j['released'] = self.released
        j['downloads'] = self.downloads
        j['direct_link'] = self.direct_link
        j['prefer_link'] = self.prefer_link
        return j

class Dependency:
    pkg_id: int = 1
    release_id: int = 1
    repo: str = REPOSITORY

    def __init__(self, pkg_id, release_id):
        self.pkg_id = pkg_id
        self.release_id = release_id
    
    @classmethod
    def create_new(cls, release):
        d = cls(
        pkg_id = release['parent_package_id'],
        release_id = release['id']
        )
        return d


    def as_json(self):
        j = json.loads("{}")
        j['pkg_id'] = self.pkg_id
        j['release_id'] = self.release_id
        return j

def next_id() -> int: 
    f = open("./assets/pkg_index.json", "r")
    j = json.loads(f.read())
    return j["next_pkg_id"]

def release_compatible(r1: str, r2: str) -> bool:
    arr1 = r1.split(".")
    arr2 = r2.split(".")
    a = (r1 == r2) or (arr1[:2] == arr2[:2])
    # print(f"{arr1} {arr2} {a}")
    return a

def sort_by_timestamp(rel):
    return rel['released']

def get_newest_release(pkg_id: int, game_version: str) -> Release:
    pkg = Package.obj_from_id(pkg_id)
    
    rels = [rel for rel in pkg.releases if release_compatible(rel['game_version'], game_version)]
    
    if len(rels) == 0 or rels == None:
        rels = pkg.releases
        print(f"compatible version for {pkg.name} mc {game_version} not found, selecting latest release")

    # print(rels)
    rels.sort(key=sort_by_timestamp)
    # print(rels)
    rels.reverse()
    return rels[0]

File: src/package.ts
import * as util from "./util"
import * as filedef from "./filedef"

export class TrackedPackage {
    cf_id: number;
    slug: string; // name of the package i.e. `xaeros-minimap`
    last_modified: number;
    constructor(cf_id: number, slug: string, last_modified: number) {
        this.cf_id = cf_id;
        this.slug = slug;
        this.last_modified = last_modified;
    }
    static mark_updated(cf_id: number, slug: string) {
        let tracked = filedef.get_tracked();
        if (!tracked.cf_ids.includes(cf_id)) {
            tracked.cf_ids.push(cf_id);
        }
        if (!tracked.packages.map( (pkg) => {return pkg.cf_id;}).includes(cf_id)) {

            let t = new this(cf_id, slug, Date.now());
            tracked.packages.push(t);

        } else {
            let ti = tracked.packages.findIndex( (pkg) => {return pkg.cf_id == cf_id});
            tracked.packages[ti].last_modified = Date.now();
        }
        filedef.write(tracked, "tracked");
    }
}

export class Release {
    id: number;
    version: string;
    game_version: string;
    deps: Array<string>; // short slugs of locators
    parent_locator: string; // short slug with the release id of this release
    released: number; // utc timestamp
    updated?: number;
    is_dependency?: boolean;
    downloads: number;
    direct_link: string;
    prefer_link: boolean;
    
    constructor(id: number, version: string, game_version: string, deps: Array<string>, parent_locator: string, released: number, is_dependency: boolean, downloads: number, direct_link: string, prefer_link: boolean) {
        this.id = id;
        this.version = version;
        this.game_version = game_version;
        this.deps = deps;
        this.parent_locator = parent_locator;
        this.released = released;
        this.is_dependency = is_dependency;
        this.downloads = downloads;
        this.direct_link = direct_link;
        this.prefer_link = prefer_link;
    }

    static create_new(version: string, game_version: string, deps: Array<string>, parent_locator: string, released: number, direct_link: string, prefer_link: boolean) {
        let idx = filedef.get_index();
        let ploc = Locator.from_short_slug(parent_locator);
        let ppkgi = idx.packages.findIndex( (pkg) => {
            return pkg.repository == ploc.repo && pkg.slug == ploc.slug;
        })
        let next_rel_id = idx.packages[ppkgi].releases.length;
        let r = new this(next_rel_id, version, game_version, deps, parent_locator, released, false, 0, direct_link, prefer_link);
        idx.packages[ppkgi].releases.push(r);
        filedef.write(idx, "index");
        return r;
    }
}

export class Package {
    name: string;
    description: string;
    releases: Array<Release>;
    repository: string;
    authors: Array<string>;
    slug: string;
    
    constructor(name: string, description: string, releases: Array<Release>, repository: string, authors: Array<string>, slug: string) {
        this.name = name;
        this.description = description;
        this.releases = releases;
        this.repository = repository;
        this.authors = authors;
        this.slug = slug;
    }

    static create_new(name: string, description: string, authors: Array<string>, releases: Array<Release>, repository: string, slug: string) 
    {
        let p = new this (name, description, releases, repository,authors, slug);
        let idx = filedef.get_index();
        idx.packages.push(p);
        idx.next_pkg_id += 1;
        filedef.write(idx, "index");
        return p;
    }

    get next_release_id() {
        return this.releases.length;
    }
}

export class Repository {
    url: string;
    api_type: number; // 1 for normal, 2 for direct
    
    constructor(url:string, api_type: number) {
        this.url = url;
        this.api_type = api_type;
    }
}

export class Locator {
    // The locator is a unique locator to a specific package version
    // It contains the full URL of the repository, so that it can be resolved even if someone hasn't added that repository
    repo: string;
    slug: string;
    rel_id: number;

    constructor(repo: string, slug: string, rel_id: number) {
        this.repo = repo
        this.slug = slug
        this.rel_id = rel_id
    }

    static from_short_slug(sslug: string) {
        let split = sslug.split("->");
        let rel_id = split.pop();
        let slug = split.pop();
        let repo = split.pop();
        return new this(repo, slug, Number(rel_id))
    }

    get short_slug() {
        return `${this.repo}->${this.slug}->${this.rel_id}`
    }
}


export function locator_to_release(locator: Locator, known_packages: Array<Package>): Release {
    known_packages = known_packages.filter( (pkg) => {return pkg.repository == locator.repo})
    if (known_packages.length == 0) {
        util.print_error(`Repository ${locator.repo} not found`);
        process.exit();
    }
    known_packages = known_packages.filter( (pkg) => {return pkg.slug == locator.slug})
    if (known_packages.length == 0) {
        util.print_error(`Package ${locator.repo}/${locator.slug} not found`);
        process.exit();
    }
    let rel: Release = undefined 
    try {
        rel = known_packages[0].releases[locator.rel_id]
    } catch (err) {
        util.print_error(`Release ${locator.short_slug} not found`)
        process.exit();
    }
    return rel
}

export function locator_to_package(locator: Locator, known_packages: Array<Package>): Package {
    known_packages = known_packages.filter( (pkg) => {return pkg.repository == locator.repo})
    if (known_packages.length == 0) {
        util.print_error(`Repository ${locator.repo} not found`);
        process.exit();
    }
    known_packages = known_packages.filter( (pkg) => {return pkg.slug == locator.slug})
    if (known_packages.length == 0) {
        util.print_error(`Package ${locator.repo}/${locator.slug} not found`);
        process.exit();
    }
    return known_packages[0]
}

export function get_desired_release(pkg: Package, game_version: string, release_version?: string): Release{
    // NOTE: Uses `latest` unless a version is specified
    let options = pkg.releases.filter( (value) => {
        return release_compatible(value.game_version, game_version);
    });
    if(release_version) {
        options = options.filter( (value) => {
            return value.version == release_version
        });
    }

    let newest = options.sort( (a, b) =>{
        return b.released - a.released;
    });
    if(newest.length === 0) {
        util.print_error(`ERROR - package ${pkg.name} for version ${game_version} not found`);
        process.exit();
    }
    return newest[0]
}

export function release_compatible(release1: string, release2: string) {
    let arr1 = release1.split(".");
    let arr2 = release2.split(".");

    return (release1 == release2 || util.arr_eq(arr1.slice(0,2), arr2.slice(0,2)))
}